csp_billing_adapter_microsoft/plugin.py

```diff
@@ -81,55 +83,74 @@ def get_account_info(config: Config):
     The information contains the metadata for compute and network.
     """
     account_info = _get_metadata()
-    account_info['compute'] = json.loads(account_info.get('compute', '{}'))
-    account_info['network'] = json.loads(account_info.get('network', '{}'))
     account_info['cloud_provider'] = get_csp_name(config)
 
     return account_info
 
 
-def get_api_header():
-    """
-    Get the header to be used in requests
-
-    Prefer IMDS which requires a token.
-    """
-    request = urllib.request.Request(
-        METADATA_TOKEN_URL,
-        headers={'Metadata': 'True'},
-        method='GET'
+def _get_metadata():
+    """Return a dict containing compute, network and signature information."""
+    metadata = {}
+    try:
+        metadata = _get_instance_metadata()
+        metadata['attestedData'] = _get_signature()
+    except ValueError as error:
+        log.error('Could not load JSON from metadata: {}'.format(error))
+        for key in ['compute', 'network', 'attestedData']:
+            if key not in metadata:
+                metadata[key] = {}
+
+    return metadata
+
+
+def _get_instance_metadata():
+    """Return all compute and network information from metadata."""
+    api_version = _get_latest_api_version()
+    if not api_version:
+        # oldest API version
+        api_version = OLDEST_API_VERSION
+
+    instance_info_url = METADATA_URL + 'instance?api-version={}'.format(
+        api_version
+    )
+    return json.loads(
+        _fetch_metadata(instance_info_url)
     )
 
-    try:
-        token = urllib.request.urlopen(request).read().decode()
-    except urllib.error.URLError as error:
-        log.error(f'Failed to retrieve metadata token: {str(error)}')
-        return {}
 
-    return token
+def _get_signature():
+    """Return attested data signature from metadata."""
+    return json.loads(
+        _fetch_metadata(SIGNATURE_URL)
```

Comment:
So here the assumption is made that `'2020-09-01'` is available but above we are prepared for now version return when fetching the available api versions. SO the code is not self consistent with the assumptions being made.